rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -83,4 +75,65 @@ Recorder::create_subscription(
   return subscription;
 }
 
+std::future<void> Recorder::launch_topics_discovery(
+  std::chrono::milliseconds topic_polling_interval,
+  bool is_discovery_disabled,
+  const std::vector<std::string> & topics_to_record)
+{
+  auto subscribe_to_topics =
+    [this, topics_to_record, topic_polling_interval, is_discovery_disabled] {
+      do {
+        auto all_topics_and_types_to_subscribe = topics_to_record.empty() ?
+          node_->get_all_topics_with_types() :
+          node_->get_topics_with_types(topics_to_record);
+
+        if (is_every_topic_subscribed(topics_to_record)) {
+          return;
+        }
+
+        subscribe_all_missing_topics(all_topics_and_types_to_subscribe);
+        std::this_thread::sleep_for(topic_polling_interval);
+      } while (rclcpp::ok() && !is_discovery_disabled);
```

Comment:
given my comment before, this condition can then be executed once on the caller side to only start the async call when discovery is enabled.

Reply:
I changed it this way, readability does not really improve imho.